### PR TITLE
kvserver: refactor replica load

### LINKED ISSUE #93579: roachtest: kv/splits/nodes=3/quiesce=true failed
roachtest.kv/splits/nodes=3/quiesce=true [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/7952891?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/7952891?buildTab=artifacts#/kv/splits/nodes=3/quiesce=true) on master @ [052acc88ad9d7296ce6b8b441627fb469cc74d95](https://github.com/cockroachdb/cockroach/commits/052acc88ad9d7296ce6b8b441627fb469cc74d95):

```
test artifacts and logs in: /artifacts/kv/splits/nodes=3/quiesce=true/run_1
(test_impl.go:291).Fatal: output in run_062628.605496985_n4_workload_run_kv: ./workload run kv --init --max-ops=1 --concurrency=192 --splits=300000 {pgurl:1-3} returned: context canceled
(test_impl.go:291).Fatal: monitor failure: monitor command failure: unexpected node event: 1: dead (exit status 137)
```
<p>Parameters: <code>ROACHTEST_cloud=gce</code>
, <code>ROACHTEST_cpu=4</code>
, <code>ROACHTEST_encrypted=false</code>
, <code>ROACHTEST_ssd=0</code>
</p>
<details><summary>Help</summary>
<p>

See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)

See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
/cc @cockroachdb/kv-triage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*kv/splits/nodes=3/quiesce=true.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

Jira issue: CRDB-22415

Epic CRDB-18656

### PR DESCRIPTION
**Pull request by kvoli**
This PR refactors the replica load stats to enforce a strict abstraction on `ReplicaLoad` and reduce the overall memory footprint.

This reduces the memory footprint of replicastats by **51%**.

running `kv/splits/nodes=3/quiesce=true` with the prior commit and this pr produces the following results when profiling heap allocated memory.

```
                              memory %
replica stats.go before       33.93
replica_stats.go after        16.60
diff (abs)                   -17.33
diff (rel)                   -51.08
```

resolves #93579
part of #87187

Release note: None

### CODE REVIEW

**Comment by kvoli** [comment]
Since this isn't urgent, going to wait on #91593 before requesting reviews here.

**Review by andrewbaptist — CHANGES_REQUESTED** [review]
> Reviewed 23 of 27 files at r1, 2 of 2 files at r2, 3 of 8 files at r3, 4 of 6 files at r4, 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/93823)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @AlexTalks and @kvoli)

___
*[`pkg/kv/kvserver/load/replica_load.go` line 132 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/93823#-NKZ6y0i14wbzxgoYp_i:-NKZ6y0i14wbzxgoYp_j:bxkevju) ([raw file](https://github.com/cockroachdb/cockroach/blob/aaf4da41aecd01251581f4c2def100596003990d/pkg/kv/kvserver/load/replica_load.go#L132)):*
> ```Go
> // replicas. See SplitRange() in store_split.go.
> func (rl *ReplicaLoad) Split(other *ReplicaLoad) {
> 	rl.mu.Lock()
> ```

Split and merge seem very deadlock prone. You discuss the need for the higher level lock. I think it is correct in the current usage, but it is likely this could be accidentally broken fairly easily. I would recommend rewriting this as

```
rl.mu.Lock()
splitStats = rl.SplitStatsAndGetAmountSplit()
rl.mu.UnLock()

other.mu.Lock()
// this only modifies other stats
other.mu.stats[i].SplitRequestCounts(splitStats)
other.mu.Unock()
```

This prevents ever having them locked simultaneously. A similar thing could be done for the merge method too. I see that you just lifted this from the other place in the code, but this still seems prefereable.

**Review by kvoli — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/93823)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @AlexTalks and @andrewbaptist)

___
*[`pkg/kv/kvserver/load/replica_load.go` line 132 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/93823#-NKZ6y0i14wbzxgoYp_i:-NKs4s148DtQ9MJPImdw:b-lo4is3) ([raw file](https://github.com/cockroachdb/cockroach/blob/aaf4da41aecd01251581f4c2def100596003990d/pkg/kv/kvserver/load/replica_load.go#L132)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

Split and merge seem very deadlock prone. You discuss the need for the higher level lock. I think it is correct in the current usage, but it is likely this could be accidentally broken fairly easily. I would recommend rewriting this as

```
rl.mu.Lock()
splitStats = rl.SplitStatsAndGetAmountSplit()
rl.mu.UnLock()

other.mu.Lock()
// this only modifies other stats
other.mu.stats[i].SplitRequestCounts(splitStats)
other.mu.Unock()
```

This prevents ever having them locked simultaneously. A similar thing could be done for the merge method too. I see that you just lifted this from the other place in the code, but this still seems prefereable.
</blockquote></details>

I don't mind this approach - however I did mention above what higher level lock is held and the function that holds it on both to prevent deadlock here.

I'd rather leave it for now and leave a TODO.

This per-replica sliding window approach isn't sustainable (especially with un-quiesced ranges by default) and these will likely move to being just counters in the near future, with a central top k tracker for a more detailed view.

**Review by andrewbaptist — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/93823#-:-NKtMiP2BiPKAkdJ25to:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @AlexTalks)

___
*[`pkg/kv/kvserver/load/replica_load.go` line 132 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/93823#-NKZ6y0i14wbzxgoYp_i:-NKtLzCe7IiZEunzh_3N:b2nhb2c) ([raw file](https://github.com/cockroachdb/cockroach/blob/aaf4da41aecd01251581f4c2def100596003990d/pkg/kv/kvserver/load/replica_load.go#L132)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

I don't mind this approach - however I did mention above what higher level lock is held and the function that holds it on both to prevent deadlock here.

I'd rather leave it for now and leave a TODO.

This per-replica sliding window approach isn't sustainable (especially with un-quiesced ranges by default) and these will likely move to being just counters in the near future, with a central top k tracker for a more detailed view.
</blockquote></details>

Ack

**Comment by kvoli** [comment]
bors r=andrewbaptist

File: pkg/kv/kvserver/replicastats/replica_stats.go [hunk 1]
```diff
@@ -89,15 +89,15 @@ type ReplicaStats struct {
 }
 
 type replicaStatsRecord struct {
-	localityCounts PerLocalityCounts
+	localityCounts *PerLocalityCounts
```
Comment (nvanbenschoten) [inline pkg/kv/kvserver/replicastats/replica_stats.go:92]: @kvoli `PerLocalityCounts` is a `map`, whose header is already a represented by a pointer. So this change is introducing an extra level of indirection and forcing this map header into a separate, 8 byte heap allocation. That's fine, but is it necessary for this change? Could we have just lazily allocated the hash map and compared the map header against nil?

See https://go.dev/play/p/mbePNNDnHji.

Comment (kvoli) [reply pkg/kv/kvserver/replicastats/replica_stats.go:92]: Certainly not necessary if the map were just left uninitialized when not used. We could have just compared the map against nil as you mentioned.

**Pull request merged**